File: week1_baseline/ruby/12_context/lib/boukensha/tools/mcp.rb
require_relative "../mcp"

module Boukensha
  module Tools
    # Public entry point for MCP-sourced tools, sitting alongside
    # Tools::FileSystem and Tools::Shell so all tool registration reads the same
    # way. The implementation lives in Boukensha::MCP.
    #
    #   Boukensha::Tools::Mcp.register(registry,
    #     command: ["kubernetes-mcp-server", "--read-only"],
    #     except:  ["pods_delete"])
    #
    #   Boukensha::Tools::Mcp.connect(registry,
    #     "mud"   => { command: [...], after_connect: "session_open" },
    #     "notes" => { command: [...] })
    module Mcp
      module_function

      # One server. See MCP::Registrar for the options.
      def register(registry, **options)
        MCP::Registrar.register(registry, **options)
      end

      # Several servers at once, namespacing any tool name they both publish.
      # See MCP::Servers. Returns [clients, provenance].
      def connect(registry, specs, debug: false)
        MCP::Servers.connect(registry, specs, debug: debug)
      end
    end
  end
end

File: week1_baseline/python/11_tui/boukensha/tools/mcp.py
"""Port of ``lib/boukensha/tools/mcp.rb``.

Public entry point for MCP-sourced tools, sitting alongside
:mod:`boukensha.tools.file_system` and :mod:`boukensha.tools.shell` so all tool
registration reads the same way. The implementation lives in
:mod:`boukensha.mcp`.

    boukensha.tools.mcp.register(
        registry,
        command=["kubernetes-mcp-server", "--read-only"],
        except_=["pods_delete"],
    )

    boukensha.tools.mcp.connect(
        registry,
        {
            "mud":   {"command": [...], "after_connect": "session_open"},
            "notes": {"command": [...]},
        },
    )
"""

from __future__ import annotations

from typing import Any

from boukensha.mcp import registrar as _registrar
from boukensha.mcp import servers as _servers
from boukensha.mcp.client import Client
from boukensha.registry import Registry


def register(registry: Registry, **options: Any) -> Client:
    """One server. See :func:`boukensha.mcp.registrar.register` for the options."""
    return _registrar.register(registry, **options)


def connect(
    registry: Registry,
    specs: dict[str, dict[str, Any]] | None,
    *,
    debug: bool = False,
) -> tuple[list[Client], list[dict[str, Any]]]:
    """Several servers at once, namespacing any tool name they both publish.

    See :func:`boukensha.mcp.servers.connect`. Returns ``(clients, provenance)``.
    """
    return _servers.connect(registry, specs, debug=debug)
